File: modules/scraml-dsl-java/src/main/java/io/atomicbits/scraml/dsl/java/client/FactoryLoader.java
/*
 *
 *  (C) Copyright 2015 Atomic BITS (http://atomicbits.io).
 *
 *  All rights reserved. This program and the accompanying materials
 *  are made available under the terms of the GNU Affero General Public License
 *  (AGPL) version 3.0 which accompanies this distribution, and is available in
 *  the LICENSE file or at http://www.gnu.org/licenses/agpl-3.0.en.html
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *  Affero General Public License for more details.
 *
 *  Contributors:
 *      Peter Rigole
 *
 */

package io.atomicbits.scraml.dsl.java.client;

/**
 * Created by peter on 10/01/16.
 */
public class FactoryLoader {

    static String defaultClientFactoryClass = "io.atomicbits.scraml.dsl.java.client.ning.Ning19ClientFactory";

    public static ClientFactory load(String clientFactoryClass) {

        String factoryClass = clientFactoryClass != null ? clientFactoryClass : defaultClientFactoryClass;

        try {
            return (ClientFactory) Class.forName(factoryClass).newInstance();
        } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
            String message = e.getMessage() + " The scraml FactoryLoader cannot load factory class " + factoryClass +
                    ". Did you add the dependency to this factory?";
            throw new NoClassDefFoundError(message);
        }
    }

}

File: modules/scraml-dsl-scala/src/main/scala/io/atomicbits/scraml/dsl/scalaplay/client/FactoryLoader.scala
/*
 *
 * (C) Copyright 2018 Atomic BITS (http://atomicbits.io).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *  Contributors:
 *      Peter Rigole
 *
 */

package io.atomicbits.scraml.dsl.scalaplay.client

import scala.language.postfixOps
import scala.util.{ Failure, Try }

/**
  * Created by peter on 8/01/16.
  */
object FactoryLoader {

  val defaultClientFactoryClass = "io.atomicbits.scraml.dsl.scalaplay.client.ning.Ning2ClientFactory"

  def load(clientFactoryClass: Option[String] = None): Try[ClientFactory] = {
    val factoryClass = clientFactoryClass.getOrElse(defaultClientFactoryClass)
    Try {
      Class.forName(factoryClass).newInstance().asInstanceOf[ClientFactory]
    } recoverWith {
      case cnfe: NoClassDefFoundError =>
        Failure(
          new NoClassDefFoundError(
            s"${cnfe.getMessage} The scraml FactoryLoader cannot load factory class $factoryClass. Did you add the dependency to this factory?"
          )
        )
      case e => Failure(e)
    }
  }

}
